From 54785cbf10629fa719fdeb7e8926041690a467ca Mon Sep 17 00:00:00 2001
From: Thiraviyam Mariyappan <quic_tmariyap@quicinc.com>
Date: Tue, 7 Feb 2023 09:52:10 +0530
Subject: [PATCH] ath12k: Add support to set tx power for 6G via iw

Add support to set the tx power for 6G radio with the wmi
pdev param WMI_PDEV_PARAM_TXPOWER_LIMIT5G via iw command.
iw dev wlanX set txpower fixed 1000
TEST=With the above iw command and verified the tx power
value in the iw dev for 6G radio is set by the configured
value.

Signed-off-by: Thiraviyam Mariyappan <quic_tmariyap@quicinc.com>
Signed-off-by: Maharaja Kennadyrajan <quic_mkenna@quicinc.com>
---
 drivers/net/wireless/ath/ath12k/core.h |  1 +
 drivers/net/wireless/ath/ath12k/mac.c  | 29 ++++++++++++--------------
 2 files changed, 14 insertions(+), 16 deletions(-)

--- a/drivers/net/wireless/ath/ath12k/core.h
+++ b/drivers/net/wireless/ath/ath12k/core.h
@@ -771,6 +771,7 @@ struct ath12k {
 	u32 max_tx_power;
 	u32 txpower_limit_2g;
 	u32 txpower_limit_5g;
+	u32 txpower_limit_6g;
 	u32 txpower_scale;
 	u32 power_scale;
 	u32 chan_tx_pwr;
--- a/drivers/net/wireless/ath/ath12k/mac.c
+++ b/drivers/net/wireless/ath/ath12k/mac.c
@@ -853,6 +853,7 @@ static void ath12k_pdev_caps_update(stru
 
 	ar->txpower_limit_2g = ar->max_tx_power;
 	ar->txpower_limit_5g = ar->max_tx_power;
+	ar->txpower_limit_6g = ar->max_tx_power;
 	ar->txpower_scale = WMI_HOST_TP_SCALE_MAX;
 }
 
@@ -905,6 +906,16 @@ static int ath12k_mac_txpower_recalc(str
 		ar->txpower_limit_5g = txpower;
 	}
 
+	if ((ar->ah->hw->wiphy->bands[NL80211_BAND_6GHZ]) &&
+		ar->txpower_limit_6g != txpower) {
+		param = WMI_PDEV_PARAM_TXPOWER_LIMIT5G;
+		ret = ath12k_wmi_pdev_set_param(ar, param,
+						txpower, ar->pdev->pdev_id);
+		if (ret)
+			goto fail;
+		ar->txpower_limit_6g = txpower;
+	}
+
 	return 0;
 
 fail:
@@ -4311,19 +4322,8 @@ static void ath12k_mac_bss_info_changed(
 	if (changed & BSS_CHANGED_TXPOWER) {
 		ath12k_dbg(ar->ab, ATH12K_DBG_MAC, "mac vdev_id %i txpower %d\n",
 			   arvif->vdev_id, info->txpower);
-
-		if (ar->supports_6ghz && info->chandef.chan &&
-                    info->chandef.chan->band == NL80211_BAND_6GHZ &&
-                    (ahvif->vdev_type == WMI_VDEV_TYPE_STA ||
-                     ahvif->vdev_type == WMI_VDEV_TYPE_AP) &&
-                    test_bit(WMI_TLV_SERVICE_EXT_TPC_REG_SUPPORT,
-                             ar->ab->wmi_ab.svc_map)) {
-                        ath12k_dbg(ar->ab, ATH12K_DBG_MAC,
-                                   "discard tx power, change to set TPC power\n");
-                } else {
-                        arvif->txpower = info->txpower;
-                        ath12k_mac_txpower_recalc(ar);
-                }
+                arvif->txpower = info->txpower;
+                ath12k_mac_txpower_recalc(ar);
 	}
 
 	if (changed & BSS_CHANGED_MCAST_RATE &&
